File: code/chapter3/linear_regression_ridge.R
library(R6)
LR_Ridge = R6Class(
  "LR_Ridge",
  public = list(
    coef = NULL,
    mu = NULL,
    sd = NULL,
    lambda = NULL,
    initialize = function(lambda) {
      self$lambda = lambda
    },
    scale = function(x) {
      self$mu = apply(x, 2, mean)
      self$sd = apply(x, 2, function(e) {
        sqrt((length(e) - 1) / length(e)) * sd(e)
      })
    },
    transform = function(x) {
      return(t((t(x) - self$mu) / self$sd))
    },
    fit = function(x, y) {
      self$scale(x)
      x_transformed = self$transform(x)
      x_lambda = rbind(x_transformed, diag(rep(sqrt(self$lambda), ncol(x))))
      y_lambda = c(y, rep(0, ncol(x)))
      self$qr_solver(cbind(c(rep(1, nrow(
        x
      )), rep(0, ncol(
        x
      ))), x_lambda), y_lambda)
    },
    qr_solver = function(x, y) {
      self$coef = qr.coef(qr(x), y)
    },
    predict = function(new_x) {
      new_x_transformed = self$transform(new_x)
      cbind(rep(1, nrow(new_x)), new_x_transformed) %*% self$coef
    }
  )
)

File: code/chapter3/linear_regression_ridge.py
import numpy as np
from sklearn.preprocessing import StandardScaler
import pdb


class LR_Ridge:
    def __init__(self, l):
        self.l = l
        self.coef = None
        self.scaler = StandardScaler()

    def qr_solver(self, x, y):
        q, r = np.linalg.qr(x)
        p = np.dot(q.T, y)
        return np.dot(np.linalg.inv(r), p)

    def fit(self, x, y):
        x_transformed = self.scaler.fit_transform(x)
        x_lambda = np.vstack(
            (x_transformed, np.diag([self.l**0.5]*x.shape[1])))
        x_lambda = np.hstack(
            (np.vstack((np.ones((x.shape[0], 1)), np.zeros((x.shape[1], 1)))), x_lambda))
        y_lambda = np.hstack((y, np.zeros((x.shape[1],))))
        self.coef = self.qr_solver(x_lambda, y_lambda)

    def predict(self, x):
        new_x_transformed = self.scaler.transform(x)
        new_x_transformed = np.hstack(
            (np.ones((x.shape[0],1)), new_x_transformed)
        )
        return np.dot(new_x_transformed, self.coef)
